grafana-plugin/pkg/plugin/proxy.go

```diff
@@ -0,0 +1,194 @@
+
+type XGrafanaContextJSONData struct {
+	ID          int    `json:"UserID"`
+	IsAnonymous bool   `json:"IsAnonymous"`
+	Name        string `json:"Name"`
+	Login       string `json:"Login"`
+	Email       string `json:"Email"`
+	Role        string `json:"Role"`
+}
+
+type OnCallProvisioningJSONData struct {
+	Error       string `json:"error,omitempty"`
+	StackId     int    `json:"stackId,omitempty"`
+	OrgId       int    `json:"orgId,omitempty"`
+	OnCallToken string `json:"onCallToken,omitempty"`
+	License     string `json:"license,omitempty"`
+}
+
+func SetXInstanceContextHeader(settings *OnCallPluginSettings, req *http.Request) error {
+	xInstanceContext := XInstanceContextJSONData{
+		StackId:      strconv.Itoa(settings.StackID),
+		OrgId:        strconv.Itoa(settings.OrgID),
+		GrafanaToken: settings.GrafanaToken,
+	}
+	xInstanceContextHeader, err := json.Marshal(xInstanceContext)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("X-Instance-Context", string(xInstanceContextHeader))
+	return nil
+}
+
+func SetXGrafanaContextHeader(user *backend.User, userID int, req *http.Request) error {
+	var xGrafanaContext XGrafanaContextJSONData
+	if user == nil {
+		xGrafanaContext = XGrafanaContextJSONData{
+			IsAnonymous: true,
+		}
+	} else {
+		xGrafanaContext = XGrafanaContextJSONData{
+			ID:          userID,
+			IsAnonymous: false,
+			Name:        user.Name,
+			Login:       user.Login,
+			Email:       user.Email,
+			Role:        user.Role,
+		}
+	}
+	xGrafanaContextHeader, err := json.Marshal(xGrafanaContext)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("X-Grafana-Context", string(xGrafanaContextHeader))
+	return nil
+}
+
+func SetAuthorizationHeader(settings *OnCallPluginSettings, req *http.Request) {
+	req.Header.Set("Authorization", settings.OnCallToken)
+}
+
+func SetOnCallUserHeader(onCallUser *OnCallUser, req *http.Request) error {
+	xOnCallUserHeader, err := json.Marshal(onCallUser)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("X-OnCall-User-Context", string(xOnCallUserHeader))
+	return nil
+}
+
+func (a *App) SetupRequestHeadersForOnCall(ctx context.Context, settings *OnCallPluginSettings, req *http.Request) error {
+	user := httpadapter.UserFromContext(ctx)
+	onCallUser, err := a.GetUser(settings, user)
+	if err != nil {
+		log.DefaultLogger.Error("Error getting user: %v", err)
+		return err
+	}
+
+	SetAuthorizationHeader(settings, req)
+
+	err = SetXInstanceContextHeader(settings, req)
+	if err != nil {
+		log.DefaultLogger.Error("Error setting instance header: %v", err)
+		return err
+	}
+
+	err = SetXGrafanaContextHeader(user, onCallUser.ID, req)
+	if err != nil {
+		log.DefaultLogger.Error("Error setting context header: %v", err)
+		return err
+	}
+
+	err = SetOnCallUserHeader(onCallUser, req)
+	if err != nil {
+		log.DefaultLogger.Error("Error setting user header: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+func (a *App) ProxyRequestToOnCall(w http.ResponseWriter, req *http.Request, pathPrefix string) {
```

Comment:
Have u considered using built in [reverse-proxy](https://go.dev/src/net/http/httputil/reverseproxy.go)? Here is the example how it's made in [grafana](https://github.com/grafana/grafana/blob/0c2f58bdaef495e4719c999882fc315e89f405a2/pkg/util/proxyutil/reverse_proxy.go#L27). Worth to check with plugin-platform team if it's something that they are using too.

Reply:
I need to close this PR in favor to https://github.com/grafana/oncall/pull/4657/files so let's discuss there